### PR TITLE
Gossip protocol implementation

### PR DESCRIPTION
**Pull request by tbekas**

### CODE REVIEW

File: modules/shared/src/main/scala/org/tesselation/ext/codecs.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+          case Left(ex)     => throw ex // FIXME: how to solve this?
+        }
+      }
+
+    implicit def decoder[F[_]: Concurrent: KryoSerializer, A <: AnyRef: ClassTag]: EntityDecoder[F, A] =
+      EntityDecoder.byteArrayDecoder[F].flatMapR { bytes =>
+        KryoSerializer[F].deserialize[A](bytes) match {
+          case Right(value) => successT[F, A](value)
+          case Left(ex) =>
+            failureT(MalformedMessageBodyFailure("Failed to deserialize http entity body with Kryo", ex.some))
```
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tesselation/ext/codecs.scala:29]: `http` -> `HTTP`

File: modules/shared/src/main/scala/org/tesselation/schema/gossip.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  case class StartGossipRoundRequest(
+    offer: List[Hash]
+  )
+
+  object StartGossipRoundRequest {
+    implicit def encoder[G[_]: KryoSerializer]: EntityEncoder[G, StartGossipRoundRequest] =
+      BinaryCodec.encoder[G, StartGossipRoundRequest]
+
+    implicit def decoder[G[_]: Concurrent: KryoSerializer]: EntityDecoder[G, StartGossipRoundRequest] =
+      BinaryCodec.decoder[G, StartGossipRoundRequest]
```
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tesselation/schema/gossip.scala:31]: I think we can create a derived macro 🤔

File: modules/shared/src/main/scala/org/tesselation/ext/codecs.scala [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.http4s.{EntityDecoder, EntityEncoder, MalformedMessageBodyFailure}
+
+object codecs {
+
+  object BinaryCodec {
+    implicit def encoder[F[_]: KryoSerializer, A <: AnyRef]: EntityEncoder[F, A] =
+      EntityEncoder.byteArrayEncoder[F].contramap { anyRef =>
+        KryoSerializer[F].serialize(anyRef) match {
+          case Right(bytes) => bytes
+          case Left(ex)     => throw ex // FIXME: how to solve this?
```
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tesselation/ext/codecs.scala:20]: Have you checked how it works? Will it throw in the runtime during encoding process? If so, it's imo fine. Kryo throws in the runtime and we can do nothing with that actually
tl;dr; as long as the error is not swallowed, it's imo fine

File: modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  rumorStorage: RumorStorage[F],
+  rumorQueue: Queue[F, RumorBatch],
+  gossip: Gossip[F]
+) extends Http4sDsl[F] {
+
+  implicit val logger = Slf4jLogger.getLogger[F]
+
+  private[routes] val prefixPath = "/gossip"
+
+  private val httpRoutes: HttpRoutes[F] = HttpRoutes.of[F] {
```
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala:28]: If the node will not end the round, then nothing will happen - no rumors will be exchanged. It's stateless, there's no state that is specific for the round.

File: modules/core/src/main/scala/org/tesselation/infrastructure/cluster/storage/RumorStorage.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  def make[F[_]: Async]: F[RumorStorage[F]] =
+    for {
+      active <- MapRef.ofConcurrentHashMap[F, Hash, Signed[Rumor]]()
+      seen <- MapRef.ofConcurrentHashMap[F, Hash, Unit]()
+      storage = new RumorStorage[F] {
+
+        val activeRetention: FiniteDuration = 2.seconds
+        val seenRetention: FiniteDuration = 2.minutes
+
+        override def addRumors(rumors: RumorBatch): F[RumorBatch] =
```
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/infrastructure/cluster/storage/RumorStorage.scala:26]: I don't agree, the API allows for fetching both rumors and hashes of rumors which should be distinguished.

File: modules/core/src/main/scala/org/tesselation/infrastructure/gossip/GossipDaemon.scala [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+    def consumeRumors: F[Unit] =
+      Stream
+        .fromQueueUnterminated(rumorQueue)
+        .evalMap { batch =>
+          batch.filterA {
+            case (hash, signedRumor) =>
+              signedRumor.value.hashF.map(_ == hash)
+          }
+        }
```
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/infrastructure/gossip/GossipDaemon.scala:55]: There is of course but we would loose the batch, and that would defeat the purpose of batching.

File: modules/core/src/main/scala/org/tesselation/infrastructure/gossip/GossipDaemon.scala [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  def make[F[_]: Async: KryoSerializer: Random: Parallel](
+    rumorStorage: RumorStorage[F],
+    rumorQueue: Queue[F, RumorBatch],
+    clusterStorage: ClusterStorage[F],
+    gossipClient: GossipClient[F],
+    rumorHandler: RumorHandler[F]
+  ): GossipDaemon[F] = new GossipDaemon[F] {
+    val fanOut: Int = 2
+    val interval: FiniteDuration = 0.2.second
+    val maxConcurrentHandlers: Int = 20
```
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/infrastructure/gossip/GossipDaemon.scala:37]: Ok 👍

File: modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala [hunk 8]
```diff
@@ -0,0 +1,54 @@
+package org.tesselation.http.routes
+
+import cats.effect.Async
+import cats.effect.std.Queue
+import cats.syntax.all._
```
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala:5]: Will do 👍

File: modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        seenHashes <- rumorStorage.getSeenHashes
+        inquiry = beginRequest.offer.diff(seenHashes)
+        beginResponse = StartGossipRoundResponse(inquiry, offer)
+        result <- Ok(beginResponse)
+      } yield result
+
+    case req @ POST -> Root / "end" =>
+      for {
+        endRequest <- req.as[EndGossipRoundRequest]
+        _ <- rumorQueue.offer(endRequest.answer)
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala:43]: exactly. both `GossipDaemon` and `GossipRoutes` should depend on the service

File: modules/core/src/main/scala/org/tesselation/infrastructure/cluster/storage/RumorStorage.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tesselation/infrastructure/cluster/storage/RumorStorage.scala:26]: it's distinguished by type

[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/infrastructure/cluster/storage/RumorStorage.scala:26]: The return type is not part of the method signature. This would be an invalid declaration:
`def get: List[Rumor]` and `def get: List[Hash]`.

File: modules/core/src/main/scala/org/tesselation/Main.scala [hunk 10]
```diff
@@ -34,48 +35,71 @@ object Main extends IOApp {
...
+                        (for {
+                          res <- AppResources.make[IO](cfg)
+                          nodeId = PeerId.fromPublic(keyPair.getPublic)
+                          _ <- Resource.eval {
+                            logger.info(s"My peer ID $nodeId")
+                          }
+                          p2pClient = P2PClient.make[IO](res.client)
+                          migrations = Migrations.make[IO]
+                          _ <- Resource.eval(migrations.migrate)
+                          queues <- Resource.eval {
```
Comment (kpudlik) [inline modules/core/src/main/scala/org/tesselation/Main.scala:54]: Imo `Queue` should be created together with implementation of Gossip in such manner:

```scala
def make[F[_]]: F[Gossip[F]] = 
  Queue.unbounded[F, ...].map(make(_))
  
def make[F[_]](queue): Gossip[F]
```

Thanks to that you encapsulate the implementation together with Resources/Queues it needs.
It also follows the pattern we have when something needs some effectful values to instantiate. Another pattern is by implicitly passing Resources but it is not the case for Queue.

File: modules/core/src/main/scala/org/tesselation/modules/Services.scala [hunk 11]
```diff
@@ -20,24 +20,28 @@ object Services {
...
+    storages: Storages[F],
+    queues: Queues[F]
   ): F[Services[F]] =
     for {
       _ <- Async[F].unit
       healthcheck = HealthCheck.make[F]
       session = Session.make[F](storages.session, storages.cluster, storages.node)
       cluster = Cluster
         .make[F](cfg, nodeId, keyPair, storages.session)
+      gossip = Gossip.make[F](queues.rumor, keyPair)
```
Comment (kpudlik) [inline modules/core/src/main/scala/org/tesselation/modules/Services.scala:32]: Imo `gossip <- Gossip.make[F](keyPair)` which already encapsulates creating of `Queue` inside and "bounds" it to the Gossip instance by returning `F[Gossip[F]]`. See my previous comment :)

File: modules/core/src/main/scala/org/tesselation/modules/Queues.scala [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+      rumorQueue <- Queue.unbounded[F, RumorBatch]
+    } yield
+      new Queues[F] {
+        override val rumor: Queue[F, RumorBatch] = rumorQueue
+      }
+}
+
+sealed abstract class Queues[F[_]] private {
+  val rumor: Queue[F, RumorBatch]
+}
```
Comment (kpudlik) [inline modules/core/src/main/scala/org/tesselation/modules/Queues.scala:22]: Imo it should be hidden inside Gossip implementation

**Review by kpudlik — CHANGES_REQUESTED** [review]

File: modules/core/src/main/scala/org/tesselation/modules/Services.scala (hunk 11 above)
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/modules/Services.scala:32]: Saw it and I disagree with it.

File: modules/core/src/main/scala/org/tesselation/domain/cluster/services/Gossip.scala [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package org.tesselation.domain.cluster.services
```
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/domain/cluster/services/Gossip.scala:1]: @marcinwadon Do we apply this also to the `Routes` and `Client`? I think that ideally, all gossip related stuff, including `Routes` and `Client` should be in the `gossip` package under `domain` and `infrastructure`.

File: modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (kpudlik) [reply modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala:43]: But `GossipRoutes` already depends on `Gossip` service and `Queue` which are passed separately 🤔 
The problem is that `GossipDaemon` does not need whole service but just the `Queue` right?
On the other hand you pass a lot of gossip-related things to `GossipDaemon` (also separately).

So I guess that discussion is mainly about treating `Gossip` as a fascade versus passing gossip-related things to services separately. Right?

[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/http/routes/GossipRoutes.scala:43]: We've talked about this offline. I am very much against this idea. It might be even impossible to do this due to the cyclic dependencies.

File: modules/core/src/main/scala/org/tesselation/infrastructure/gossip/GossipDaemon.scala [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+    private val logger = Slf4jLogger.getLogger[F]
+
+    override def startDaemon: F[Unit] =
+      for {
+        _ <- Spawn[F].start(spreadActiveRumors.foreverM).void
+        _ <- Spawn[F].start(consumeRumors)
+      } yield ()
```
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/infrastructure/gossip/GossipDaemon.scala:45]: When I stopped the process with SIGINT (Ctrl+C) it closed normally, it didn't interrupt the shutdown of the process, meaning it had to cancel.

File: modules/core/src/main/scala/org/tesselation/domain/cluster/services/Gossip.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/domain/cluster/services/Gossip.scala]: `Gossip` is a part of `Services` in the same way that `GossipClient` is a part of `P2PClietnt`. I think that the same logic should be applied to both of these classes.

File: modules/core/src/main/scala/org/tesselation/config/types.scala [hunk 15]
```diff
@@ -34,6 +35,22 @@ object types {
     idleTimeInPool: FiniteDuration
   )
 
+  case class RumorStorageConfig(
+    activeRetention: FiniteDuration,
+    seenRetention: FiniteDuration
+  )
+
+  case class GossipDaemonConfig(
+    fanOut: Int,
```
[the remark this answers is not in the document]
Comment (tbekas) [reply modules/core/src/main/scala/org/tesselation/config/types.scala:44]: Ok, then I will wait for your change.

**Review by marcinwadon — APPROVED** [review]
